### PR TITLE
Consume .editorconfig options for standard .editorconfig values

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This brings in core support for reading in .editorconfig files for "standard" options as defined [here](http://editorconfig.org/#file-format-details). This also includes a few other fixups and cleanups from my previous options pull request, so I'd suggest you do this a commit at a time.

*Review:* @dotnet/roslyn-ide

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please.

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs [hunk 1]
```diff
@@ -1,23 +1,39 @@
...
 
         // This is also serialized by the Visual Studio-specific LanguageSettingsPersister
-        public static PerLanguageOption<int> TabSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(TabSize), defaultValue: 4);
+        public static PerLanguageOption<int> TabSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(TabSize), defaultValue: 4,
+            storageLocations: new EditorConfigStorageLocation("tab_width", s => int.Parse(s)));
 
         // This is also serialized by the Visual Studio-specific LanguageSettingsPersister
-        public static PerLanguageOption<int> IndentationSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(IndentationSize), defaultValue: 4);
+        public static PerLanguageOption<int> IndentationSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(IndentationSize), defaultValue: 4,
+            storageLocations: new EditorConfigStorageLocation("indent_size", s => int.Parse(s)));
```
Comment (rchande) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: What happens if this fails?

Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: Uh, right now, badness. :-) D'oh.

File: src/EditorFeatures/Core/Implementation/TodoComment/TodoCommentTokens.cs [hunk 2]
```diff
@@ -73,10 +73,9 @@ public TokenInfo(string optionText, ImmutableArray<TodoCommentDescriptor> tokens
 
         private TokenInfo _lastTokenInfo;
 
-        public async Task<ImmutableArray<TodoCommentDescriptor>> GetTokensAsync(Document document, CancellationToken cancellationToken)
+        public ImmutableArray<TodoCommentDescriptor> GetTokens(Document document, CancellationToken cancellationToken)
         {
-            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
-            var optionText = documentOptions.GetOption(TodoCommentOptions.TokenList);
+            var optionText = document.Project.Solution.Options.GetOption(TodoCommentOptions.TokenList);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/TodoComment/TodoCommentTokens.cs]: so this is basically stating that TODO comments aren't configurable through .editor_config?

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    // NOTE: this type depends on Dev15 assemblies, which is why the type is in EditorFeatures.Next. But, that library
+    // is rehostable and once we move .editorconfig support fully through the system, it should be moved to Workspaces
+    // or perhaps even lower.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: why not make a Workspaces.Next?

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    internal sealed class EditorConfigDocumentOptionsProvider : IDocumentOptionsProvider
+    {
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// The map of cached contexts for currently open documents.
+        /// </summary>
+        private readonly Dictionary<DocumentId, Task<ICodingConventionContext>> _openDocumentContexts = new Dictionary<DocumentId, Task<ICodingConventionContext>>();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: i assuem, because of the _gate, that this will be accesed from different threads.  can you doc as much?

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            lock (_gate)
+            {
+                Task<ICodingConventionContext> contextTask;
+
+                if (_openDocumentContexts.TryGetValue(e.Document.Id, out contextTask))
+                {
+                    _openDocumentContexts.Remove(e.Document.Id);
+
+                    // Ensure we dispose the context, which we'll do asynchronously
+                    contextTask.ContinueWith(t => t.Result.Dispose(), CancellationToken.None, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: wrap.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: also, any desire to cancel the task that is executing?

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            lock (_gate)
+            {
+                _openDocumentContexts.TryGetValue(document.Id, out contextTask);
+            }
+
+            if (contextTask != null)
+            {
+                // The file is open, let's reuse our cached data for that file. The task might still be running, but we want to allow for eager cancellation
+                // if the caller doesn't need it.
+                var cancellableContextTask = contextTask.ContinueWith(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: the comment isn't really clear.  I think you want to explain that 'awaiting' the contextTask means that we won't respect 
cancellation.  So we do this so we can both get the underlying value and respect cancellation if the caller wants that.

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                if (path == null)
+                {
+                    if (document.Name != null && document.Project.FilePath != null)
+                    {
+                        path = Path.Combine(Path.GetDirectoryName(document.Project.FilePath), document.Name);
+                    }
+                    else
+                    {
+                        // Really no idea where this is going, so bail
+                        return null;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: have we doc'ed the interface to make it clear this is a possibility?  I would not have expected it.

File: src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs [hunk 8]
```diff
@@ -56,9 +56,12 @@ internal sealed class LanguageSettingsPersister : ForegroundThreadAffinitizedObj
             {
                 var languagePreferences = new LANGPREFERENCES3[1];
                 languagePreferences[0].guidLang = languageGuid;
-                Marshal.ThrowExceptionForHR(_textManager.GetUserPreferences4(pViewPrefs: null, pLangPrefs: languagePreferences, pColorPrefs: null));
 
-                RefreshLanguageSettings(languagePreferences);
+                // The function can potentially fail if that language service isn't installed
+                if (ErrorHandler.Succeeded(_textManager.GetUserPreferences4(pViewPrefs: null, pLangPrefs: languagePreferences, pColorPrefs: null)))
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs]: if possible, can we cherry pick this out and get in soon.

File: src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs [hunk 9]
```diff
@@ -56,9 +56,12 @@ internal sealed class LanguageSettingsPersister : ForegroundThreadAffinitizedObj
             {
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs]: regarding not knowing about TypeScript's guid.  Why can't we have AbstractPackage (or whatever) register some hook into this.  That way we'll know about all these langauges and guids in a dynamic fashion instead of needing to hardcode it in?

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs [hunk 10]
```diff
@@ -1,23 +1,39 @@
...
+using System;
 using Microsoft.CodeAnalysis.Options;
 
 namespace Microsoft.CodeAnalysis.Formatting
 {
     public static class FormattingOptions
     {
-        public static PerLanguageOption<bool> UseTabs { get; } = new PerLanguageOption<bool>(nameof(FormattingOptions), nameof(UseTabs), defaultValue: false);
+        public static PerLanguageOption<bool> UseTabs { get; } = new PerLanguageOption<bool>(nameof(FormattingOptions), nameof(UseTabs), defaultValue: false,
+            storageLocations: new EditorConfigStorageLocation("indent_style", s => s == "tab"));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: what's the intuition on these values? are they basically waht we expect the actual .editorConfig file to contain?

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs [hunk 11]
```diff
@@ -1,23 +1,39 @@
...
-        public static PerLanguageOption<int> TabSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(TabSize), defaultValue: 4);
+        // This is also serialized by the Visual Studio-specific LanguageSettingsPersister
+        public static PerLanguageOption<int> TabSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(TabSize), defaultValue: 4,
+            storageLocations: new EditorConfigStorageLocation("tab_width", s => int.Parse(s)));
 
-        // This is also serialized by the Visual Studio-specific LanguageSettingsSerializer
-        public static PerLanguageOption<int> IndentationSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(IndentationSize), defaultValue: 4);
+        // This is also serialized by the Visual Studio-specific LanguageSettingsPersister
+        public static PerLanguageOption<int> IndentationSize { get; } = new PerLanguageOption<int>(nameof(FormattingOptions), nameof(IndentationSize), defaultValue: 4,
+            storageLocations: new EditorConfigStorageLocation("indent_size", s => int.Parse(s)));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: what if this function throws an exception?

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: tecnically, you can just pass int.Parse.

File: src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Microsoft.CodeAnalysis.Options
+{
+    /// <summary>
+    /// Specifies that an option should be read from an .editorconfig file.
+    /// </summary>
+    internal sealed class EditorConfigStorageLocation : OptionStorageLocation
+    {
+        public string KeyName { get; }
+        public Func<string, object> ParseFunction { get; }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs]: 'ParseValue' perhaps?

Possibly hide that this is a Func, and just have a method you expose that calls the func?

File: src/Workspaces/Core/Portable/Options/IDocumentOptions.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Options
+{
+    /// <summary>
+    /// Returned from a <see cref="IDocumentOptionsProvider"/>
+    /// </summary>
+    interface IDocumentOptions
+    {
+        bool TryGetDocumentOption(Document document, OptionKey option, out object value);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Options/IDocumentOptions.cs]: Do we want a ```<T>``` version?

File: src/Workspaces/Core/Portable/Options/IOptionService.cs [hunk 14]
```diff
@@ -46,5 +48,8 @@ internal interface IOptionService : IWorkspaceService
         IEnumerable<IOption> GetRegisteredOptions();
 
         event EventHandler<OptionChangedEventArgs> OptionChanged;
+
+        void RegisterDocumentOptionsProvider(IDocumentOptionsProvider documentOptionsProvider);
+        Task<OptionSet> GetAmendedOptionSetForDocumentAsync(Document document, OptionSet optionSet, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Options/IOptionService.cs]: what does 'Amended' mean here?

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 15]
```diff
@@ -456,16 +456,30 @@ private string GetDebuggerDisplay()
...
         /// </summary>
         /// <remarks>
-        /// This method is async because this may require reading other files. It is expected this is cheap and synchronous in most cases due to caching.</remarks>
+        /// This method is async because this may require reading other files. In files that are already open, this is expected to be cheap and complete synchronously.
+        /// </remarks>
         public Task<DocumentOptionSet> GetOptionsAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
-            // TODO: merge with document-specific options
-            return Task.FromResult(new DocumentOptionSet(Project.Solution.Options, Project.Language));
+            if (_cachedOptions == null)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: do we really care about being lazy about created _cachedOptions.  Why not just make the AsyncLazy.  It won't actually compute anything until someone pulls on it right?

File: src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs (hunk 8 above)
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs]: +1 on getting this cherry picked sooner. I am currently hitting this exception when trying to run VSI tests on Willow install which doesn't have TS language service.

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs (hunk 10 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: Yes, this is decided by whoever figured out the .editorconfig spec. In this particular case, these are well-defined values from the core spec, so we're obviously using them.

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs (hunk 15 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: We still would be creating an AsyncLazy being created for each Document, with (in-VS) at least only a minority of them actually being used. This was me optimizing (prematurely, I admit) to avoid those allocations.

File: src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs (hunk 8 above)
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs]: I've just merged #13869 to fix that.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test this please. (Just using this as a test for #9222 being fixed.)

**Comment by jasonmalinowski** [comment]
@dotnet-bot test this please. (Verifying #9222.)

**Comment by jasonmalinowski** [comment]
@dotnet-bot test this please. (Verifying #9222.)

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs (hunk 3 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: Laziness, since eventually such a split would go away once we finish the move of this down to the compiler/workspaces layer.

File: src/Workspaces/Core/Portable/Options/IDocumentOptions.cs (hunk 13 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Options/IDocumentOptions.cs]: This is only called by the core options infrastructure, at the point it's no longer generic anyways. It doesn't change anything.

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs (hunk 7 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: Fixing.

File: src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs (hunk 9 above)
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs]: That's basically what we'll do, yes.

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs (hunk 6 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: Updating.

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs (hunk 5 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: It seemed like some unnecessary complexity, also strangeness could then begin if the task we're awaiting later then gets that cancelled out underneath it. Yeah, we could use AsyncLazy and whatnot but I don't anticipate a problem here.

File: src/EditorFeatures/Core/Implementation/TodoComment/TodoCommentTokens.cs (hunk 2 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/TodoComment/TodoCommentTokens.cs]: For now, yes. Once we figure out a way to represent that (and we've probably moved that provider out of proc for performance), we can put this back.

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs (hunk 4 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: Fixed.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please, integration tests are unblocked.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please. Fixed test setup issue.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please. Fixed deadlock.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please. Validating with future-proof fix.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test windows_vsi_p3_open_prtest please. This got hit by the ErrorList bug.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please. Updated groovy scripts to correctly take logs.

**Comment by jmarolf** [comment]
test vsi please

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please.

Testing with some instrumentation.

File: src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs [hunk 16]
```diff
@@ -134,6 +134,7 @@ public bool TryGetDocumentOption(Document document, OptionKey option, out object
 
                 if (_codingConventionSnapshot.TryGetConventionValue(editorConfigPersistence.KeyName, out value))
                 {
+                    FailFast.OnFatalException(new Exception("we actually got an .editorconfig value?"));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs]: If this passes please don't merge this in.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test vsi please. Had to resolve some merge conflicts.

**Comment by jasonmalinowski** [comment]
And the test issues have been addressed (and GitHub is actually up) so tagging @MattGertz and @Pilchie to confirm approval to merge this for RC.

File: src/EditorFeatures/Next/EditorFeatures.Next.csproj [hunk 17]
```diff
@@ -54,6 +54,7 @@
     <Reference Include="System" />
     <Reference Include="System.ComponentModel.Composition" />
     <Reference Include="System.Core" />
+    <Reference Include="$(NuGetPackageRoot)\VS.ExternalAPIs.CodingConventions\1.0.60704.2\Microsoft.VisualStudio.CodingConventions.dll" />
```
Comment (Pilchie) [inline src/EditorFeatures/Next/EditorFeatures.Next.csproj]: Why is this here?  Isn't it in the project.json below?

Comment (jasonmalinowski) [inline src/EditorFeatures/Next/EditorFeatures.Next.csproj]: The binary isn't in lib so NuGet doesn't process it correctly.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_release_unit32_prtest  please. This hit #13980 which is fixed in master but not dev15-rc.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_release_unit32_prtest please. This hit #13980 which is fixed in master but not dev15-rc.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_release_unit32_prtest please. Perhaps the SolutionExplorerShim issue is my fault, but the artifacts from the build have already expired.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_release_unit32_prtest please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest this please.
